---
title: "note"
author: "navych"
date: "2022/9/22"
output: html_document

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#knitr::opts_knit$set(root.dir = '..')
```

## Generate TSS annotation
```{r ,warning=FALSE,message=FALSE}
library(dplyr)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
transc<-transcripts(TxDb.Hsapiens.UCSC.hg38.knownGene)
tss<-data.frame(chrom=as.vector(transc@seqnames),tss=transc@ranges@start-1,end=transc@ranges@start,strand=transc@strand)
data<-filter(tss, chrom%in%paste0('chr',c(1:22,'X','Y')))
write.table(data, '../data/annotation/tss.bed',row.names = FALSE,sep = '\t',quote = FALSE, col.names = FALSE)
```

## Generate TSS signal Matrix
```{r}
system("python ./script/methTssSignalMatrix.py -t ./data/annotation/tss.bed -i ./data/LAD.group.ratio.bed -u 15000 -d 15000 -o ./data/intermediate/signalMatrixGroup.tsv")
```

## Find SR-DMRs from SR-DMCs
```{r}
setwd('..')
system("python ./script/dmc2dmr.py -i ./data/intermediate/srdmc.s2.bed -o ./data/intermediate/srdmr.s2.bed")

```
## Get the Methylation Level of SR-DMC and SR-DMR
```{bash}

python ./script/methMatrixExtract.py -b /path/to/merge_ratio_bed.gz -r ./data/intermediate/srdmc.s3.bed -o ./data/intermediate/srdmc.s3.ratio.bed
python ./script/methMatrixExtract.py -b /path/to/merge_ratio_bed.gz -r ./data/intermediate/srdmr.s3.bed -o ./data/intermediate/srdmr.s3.ratio.bed
```

# Supplementary
## Genomic element annotation
The annotation of GpG islands (CGI) is download from [UCSC Table Browser](https://genome.ucsc.edu/cgi-bin/hgTables?hgsid=1773909416_i3FB69HY3DMQvl1e7YgCNw7Iqc3r).

